game/src/api/utils/DelayedSet.java

```diff
@@ -0,0 +1,94 @@
+import java.util.ConcurrentModificationException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class allows managing a Collection inside the System-Loops.
+ *
+ * <p>Since the Systems iterate over a Set of Entity's, a {@link ConcurrentModificationException} is
+ * normally thrown if this System adds/removes an entity of this set.
+ *
+ * <p>This class implements three different HashSets to allow delayed manipulation of the base Set.
+ *
+ * <p>{@link #current} contains the current "active" object of the collection. Use this Set to
+ * iterate over and work with. Use {@link #getSet()} to get this Set.
+ *
+ * <p>Use {@link #add} to add the given object to {@link #toAdd}. After the call of {@link #update},
+ * the objects inside this inner set will be added to {@link #current}
+ *
+ * <p>Use {@link #remove} to add the given object to {@link #toRemove}. After the call of {@link
+ * #update}, the objects inside this inner set will be removed from {@link #current}
+ *
+ * <p>Use {@link #update()} to add/remove the objects to/from {@link #current}. If you do this
+ * inside of an iteration over this Set, a {@link ConcurrentModificationException} will be thrown.
+ * It is best to use this right before or right after iterating over the set.
+ *
+ * @param <T> Type of the Elements to store in the Sets.
+ * @see ConcurrentModificationException
+ * @see HashSet
+ */
+public final class DelayedSet<T> {
+
+    private final Set<T> current = new HashSet<>();
+    private final Set<T> toAdd = new HashSet<>();
+    private final Set<T> toRemove = new HashSet<>();
+
+    /**
+     * Update the {@link #current} based on the elements in {@link #toAdd} and {@link #toRemove}.
+     *
+     * <p>Add all objects from {@link #toAdd} to {@link #current}. Remove all objects from {@link
+     * #toRemove} to {@link #current}. Clears {@link #toAdd} and {@link #toAdd}
+     */
+    public void update() {
+        current.addAll(toAdd);
+        current.removeAll(toRemove);
+        toAdd.clear();
+        toRemove.clear();
+    }
+
+    /**
+     * Add the given to {@link #toAdd}.
+     *
+     * @param t Object to add
+     */
+    public void add(T t) {
+        toAdd.add(t);
+    }
+
+    /**
+     * Add all objects of the given collection to {@link #toAdd}.
+     *
+     * @param collection contains all objects to add
+     */
+    public void addAll(Collection<T> collection) {
+        toAdd.addAll(collection);
+    }
+
+    /**
+     * Add the given to {@link #toRemove}.
+     *
+     * @param t Object to remove
+     */
+    public void remove(T t) {
+        toRemove.add(t);
+    }
+
+    /**
+     * Add all objects of the given collection to {@link #toRemove}.
+     *
+     * @param collection contains all objects to remove
+     */
+    public void removeAll(Collection<T> collection) {
+        toRemove.addAll(collection);
+    }
+
+    /**
+     * @return The Set with all currently active elements
+     */
+    public Set<T> getSet() {
```

Comment:
Hier bin ich etwas unglücklich. Du reichst direkt eine Referenz auf die interne Repräsentation heraus und erlaubst damit die Manipulation von außen. Spotbugs würde das (zu Recht) anmeckern.

Sollen wir das Interface aus `Set` nehmen: `Object[] toArray()`?

Und möglicherweise brauchen wir auch noch `Iterator<E> iterator()`, damit das Ding wie ein `Set` bearbeitet werden kann? Aber da wir das Interface `Set` nicht implementieren (können/wollen), ist das nur für die Nutzung in Schleifen relevant.

Reply:
> Sollen wir das Interface aus Set nehmen: Object[] toArray()?

Hmm, das ist zum handeln aber echt mäh... Dann hast du a) außen überall ein Array und keine Collection und b) alles ist von Typ `Object` und du musst es erst Parsen (und theoretisch abfangen, wenn es nicht geht)

Reply:
> Du hättest ja direkt von einer konkreten Set ableiten können, beispielsweise von der, die wir sowie nutzen, und dann nur die add/remove-Methoden überschreiben müssen?

Fürs Protokoll, ich hab das eben versucht (ableiten von `HashSet`) und bin nicht mehr auf einen grünen Ast gekommen. Neben einer Vielzahl an Signaturen die mir aufgehängt wurden (z,B`boolean remove (Object)`), sind mir auch mehrere Methoden über den weg gelaufen, die irgendwie mit add/remove zu tun haben Teilweise aber nicht auf mein `add`/`remove` zugegriffen haben. 
Hab irgendwann den Stecker gezogen. Habs also nicht bis zum Ende durchgetüftelt.

Reply:
> Hmm, das ist zum handeln aber echt mäh

true that. 

> Könnte ich nicht einfach eine Kopie des Sets rausgeben?

Damit würdest Du zumindest nicht die interne Struktur "offenbaren". 

Ich hab grad nochmal im Interfaces zu [`Set`](https://docs.oracle.com/javase/tutorial/collections/interfaces/set.html) und [`Collection`](https://docs.oracle.com/javase/tutorial/collections/interfaces/collection.html) geschaut - sowas in der Art könnten die doch auch gebrauchen. Haben sie aber aus irgendeinem Grund nicht. Vermutlich wird das über die Methode `iterator()` und/oder die Methode `stream()` abgehandelt, was ich eigentlich relativ elegant finde - das Ding kannst Du quasi überall in Schleifen-Kontexten gut nutzen.

Reply:
> > Du hättest ja direkt von einer konkreten Set ableiten können, beispielsweise von der, die wir sowie nutzen, und dann nur die add/remove-Methoden überschreiben müssen?
> 
> Fürs Protokoll, ich hab das eben versucht (ableiten von `HashSet`) und bin nicht mehr auf einen grünen Ast gekommen. Neben einer Vielzahl an Signaturen die mir aufgehängt wurden (z,B`boolean remove (Object)`), sind mir auch mehrere Methoden über den weg gelaufen, die irgendwie mit add/remove zu tun haben Teilweise aber nicht auf mein `add`/`remove` zugegriffen haben. Hab irgendwann den Stecker gezogen. Habs also nicht bis zum Ende durchgetüftelt.

true that. ich habe den gedanken vorhin nicht zuende gedacht - durch die "verzögerung" bei `add` und `remove` müssten dann _alle_ methoden aus `HashSet` modifiziert werden, die irgendwas entfernt mit hinzufügen/löschen zu tun haben. also ist dein weg schon besser.